File: server/src/atelier/api/plugins.py
"""插件管理接口：列插件、触发安装、查安装进度。

安装是后台线程干的活（见 `plugins.py`），这里三个端点都秒回：install 只是把线程点着，
真进度靠前端轮询 GET 拿。状态存进程内存，不碰任何库。
"""

from __future__ import annotations

from fastapi import APIRouter, HTTPException, status
from pydantic import BaseModel

from atelier import plugins

router = APIRouter(prefix="/api/plugins", tags=["plugins"])


class PluginOut(BaseModel):
    id: str
    name: str
    description: str
    installed: bool
    running: bool
    progress: int
    eta_seconds: int | None = None
    downloaded_bytes: int = 0
    total_bytes: int = 0
    speed_bytes: int | None = None
    message: str


@router.get("", response_model=list[PluginOut])
def list_all() -> list[PluginOut]:
    return [PluginOut(**one) for one in plugins.list_plugins()]


@router.get("/{plugin_id}", response_model=PluginOut)
def read_one(plugin_id: str) -> PluginOut:
    try:
        return PluginOut(**plugins.plugin_status(plugin_id))
    except KeyError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=f"没有这个插件：{plugin_id}"
        ) from exc


@router.post("/{plugin_id}/install", response_model=PluginOut)
def install(plugin_id: str) -> PluginOut:
    try:
        return PluginOut(**plugins.start_install(plugin_id))
    except KeyError as exc:
        raise HTTPException(
            status_code=status.HTTP_404_NOT_FOUND, detail=f"没有这个插件：{plugin_id}"
        ) from exc

File: app/src/api/plugins.ts
/**
 * 插件管理接口：列插件、触发安装、查进度。
 *
 * 安装是后台线程干的，install 秒回；真进度靠页面轮询 `getPlugin` 拿。
 */
import { request } from './client'

export interface Plugin {
  id: string
  name: string
  description: string
  installed: boolean
  running: boolean
  /** 0-100，安装中才有意义。 */
  progress: number
  /** 预计剩余秒数，安装中才有；刚开始测不到速率时为 null。 */
  eta_seconds: number | null
  /** 已下字节，安装中/刚完成才有意义。 */
  downloaded_bytes: number
  /** 总字节，列文件拿到大小后才 > 0。 */
  total_bytes: number
  /** 当前下载速率（字节/秒），测不到时为 null。 */
  speed_bytes: number | null
  /** 失败时的原因，正常为空串。 */
  message: string
}

export function listPlugins(): Promise<Plugin[]> {
  return request<Plugin[]>('/api/plugins')
}

export function getPlugin(id: string): Promise<Plugin> {
  return request<Plugin>(`/api/plugins/${encodeURIComponent(id)}`)
}

export function installPlugin(id: string): Promise<Plugin> {
  return request<Plugin>(`/api/plugins/${encodeURIComponent(id)}/install`, { method: 'POST' })
}
